plutus-tx/src/PlutusTx/Data.hs

```diff
@@ -90,34 +88,91 @@ the indefinite kinds.
 -}
 
 -- | Turn a CBOR Term into Data if possible.
-fromTerm :: CBOR.Term -> Either String Data
-fromTerm = \case
-    -- See Note [CBOR alternative tags]
-    CBOR.TTagged t (CBOR.TList ds)
-        | 121  <= t && t < 128                -> Constr (fromIntegral t - 121) <$> traverse fromTerm ds
-    CBOR.TTagged t (CBOR.TList ds)
-        | 1280 <= t && t < 1401               -> Constr ((fromIntegral t - 1280) + 7) <$> traverse fromTerm ds
-    CBOR.TTagged t (CBOR.TList (i:ds))
-        | t == 102, Just actualTag <- asInt i -> Constr actualTag <$> traverse fromTerm ds
-    CBOR.TTagged _ _                          -> throwError "Couldn't interpret tag as constructor tag"
-    -- See Note [Definite and indefinite forms of CBOR]
-    CBOR.TMap es                              -> Map <$> traverse (\(t1, t2) -> (,) <$> fromTerm t1 <*> fromTerm t2) es
-    CBOR.TMapI es                             -> Map <$> traverse (\(t1, t2) -> (,) <$> fromTerm t1 <*> fromTerm t2) es
-    CBOR.TList l                              -> List <$> traverse fromTerm l
-    CBOR.TListI l                             -> List <$> traverse fromTerm l
-    CBOR.TInteger i                           -> pure $ I i
-    CBOR.TInt i                               -> pure $ I $ fromIntegral i
-    CBOR.TBytes b                             -> if BS.length b <= 64
-                                                   then pure $ B b
-                                                   else throwError "ByteString exceeds 64"
-    CBOR.TBytesI b                            -> if BSL.length b <= 64
-                                                   then pure $ B $ BSL.toStrict b
-                                                   else throwError "ByteString exceeds 64"
-    _                                         -> throwError "Unsupported kind of CBOR"
-
--- See Note [Definite and indefinite forms of CBOR]
--- | View a CBOR Term as an Integer if possible.
-asInt :: CBOR.Term -> Maybe Integer
-asInt (CBOR.TInteger i) = Just i
-asInt (CBOR.TInt i)     = Just $ fromIntegral i
-asInt _                 = Nothing
+decodeData :: forall s. Decoder s Data
+decodeData = CBOR.peekTokenType >>= \case
+  CBOR.TypeUInt         -> decodeInt
+  CBOR.TypeUInt64       -> decodeInt
+  CBOR.TypeNInt         -> decodeInt
+  CBOR.TypeNInt64       -> decodeInt
+  CBOR.TypeInteger      -> decodeBigInt
+
+  CBOR.TypeBytes        -> decodeBytes
+  CBOR.TypeBytesIndef   -> decodeBytes
+
+  CBOR.TypeListLen      -> decodeList
+  CBOR.TypeListLen64    -> decodeList
+  CBOR.TypeListLenIndef -> decodeList
+
+  CBOR.TypeMapLen       -> decodeMap
+  CBOR.TypeMapLen64     -> decodeMap
+  CBOR.TypeMapLenIndef  -> decodeMap
+
+  CBOR.TypeTag          -> decodeConstr
+  CBOR.TypeTag64        -> decodeConstr
+
+  _                     -> fail "Unrecognized value"
+
+decodeInt :: Decoder s Data
+decodeInt = I <$> CBOR.decodeInteger
+
+decodeBigInt :: Decoder s Data
+decodeBigInt = do
+  convert <- CBOR.decodeTag64 >>= \case
```

Comment:
Can we not use https://hackage.haskell.org/package/cborg-0.2.5.0/docs/Codec-CBOR-Decoding.html#v:decodeInteger ?

Reply:
Is it so that you can put the check in the middle? It seems like it might be simpler to reuse decoders for integers and bytestrings and then do the size check afterwards?

Reply:
The size check for the bytestring is in terms of length, but for an integer it's less direct. We'd either hard code in a really big number or do something with logs or something.

Reply:
Couldn't we check the bit-size of the Integer?

Reply:
How does one do that?

Reply:
https://hackage.haskell.org/package/base-4.15.0.0/docs/Data-Bits.html#v:popCount, I think.

Reply:
My read of that function is that the popcount of `2^12345` is 1.

Reply:
Yes, you're right. I'm sure there should be a way to use that module to get the highest set bit...

Reply:
> countLeadingZeros can be used to compute log base 2 via
> logBase2 x = finiteBitSize x - 1 - countLeadingZeros x

seems dumb but :shrug: 

Of course, this would rely on us making some assumptions that integers of a given size are represented with a particular amount of bytes, which may or may not be true, I guess.

Reply:
I don't doubt that there's a way to do it.

But the method used would measure the number of bytes indirectly through some arithmetic, which obscures the intent of the code.

Reply:
Sure, but the current code requires us to re-implement a somewhat scary-looking decoder for integers which relies on `CBOR.Magic` etc., rather than using an existing one from somewhere else. _That_ obscures the intent of the code if anything!

Anyway, I won't die on this hill, probably we won't touch this much in future anyway.

Reply:
It is a rather sketchy dependency, isn't it.

Would you prefer a version using this along with some hard-coded number? https://hackage.haskell.org/package/integer-gmp-1.0.2.0/docs/GHC-Integer-Logarithms.html#v:integerLog2-35-

Reply:
I'm happy with it if you believe we can make a true argument about the size of the underlying bytes based on that.

Reply:
We'd ultimately be replacing the uncomfortable dependency on an internal module with an uncomfortable argument for correctness based on a spec and some math.

Reply:
I guess the degree of discomfort could vary.

Reply:
There might be more options here. I think I need to stew on this a bit.

Reply:
How do you do this in the ledger? Presumably you also have to worry about not decoding big integers? Do you just only have Int64s?

Perhaps we could put some `decodeNotTooBigX` functions into `cardano-base` and use them from both?

Reply:
There's always this:
```
max :: Integer
max = 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff

inBounds :: Integer -> Bool
inBounds x = (x <= max) && (x >= -1 - max)
```

Reply:
Ledger values are bounded by int64 at most

Reply:
Well, technically just outside of in64 range.